Dynamic alert condition config for pluggable alert conditions

dennisoelkers opened the pull request:
## Description

This change makes it possible to create custom AlertConditions which are loadable from plugins and  are either configured using the standard configuration forms (similar to inputs, outputs, alarm callbacks etc.) or provide their own configuration form components. Same for the summary of the alert conditions. Custom form and summary components are used for the builtin alert conditions.

Refs #2867.

## Motivation and Context
This change is part of the enhanced alerting story, which allows users to provide their own customized alert conditions if the builtin ones are not sufficient for their specific use case.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

joschi commented on graylog2-server/src/main/java/org/graylog2/alerts/AlertServiceImpl.java:
```diff
@@ -158,7 +158,7 @@ private AlertCondition createAlertCondition(String type,
     @Override
     public AlertCondition fromRequest(CreateConditionRequest ccr, Stream stream, String userId) {
         final String type = ccr.type();
-        checkArgument(type != null, "Milling alert condition type");
```
![Windmill](http://i.giphy.com/kpqqXxSAkLK48.gif)

joschi commented on graylog2-server/src/main/java/org/graylog2/alerts/types/FieldContentValueAlertCondition.java:
```diff
@@ -61,6 +64,33 @@ FieldContentValueAlertCondition create(Stream stream,
+        public Descriptor() {
+            super(
+                "Field Content Value Alert Condition",
+                "http://www.graylog.rog",
```
Typo, should be `https://www.graylog.org/`.

joschi commented on graylog2-server/src/main/java/org/graylog2/alerts/types/FieldContentValueAlertCondition.java:
```diff
@@ -61,6 +64,33 @@ FieldContentValueAlertCondition create(Stream stream,
+                "Field Content Value Alert Condition",
+                "http://www.graylog.rog",
+                "This condition is triggered when the value of a defined field in a defined time interval has a "
+                + "mean/mininum/maximum value/standard deviation which is higher/lower than a defined threshold."
```
Can we simplify the second half of this sentence? It has quite a few "options".

dennisoelkers replied:
Maybe "This condition is triggered when the aggregated value of a field is higher/lower than a defined threshold for a given time range."?

joschi replied:
Yes, much better! 👍

dennisoelkers replied:
✅

joschi commented on graylog2-server/src/main/java/org/graylog2/alerts/types/MessageCountAlertCondition.java:
```diff
@@ -26,28 +26,46 @@
+    enum ThresholdType {
+
+        MORE("more than"),
+        LESS("less than");
```
Should there also be an `EXACT` threshold type (e. g. exactly 10 messages) or is this subject to a future feature (if at all)?

dennisoelkers replied:
Hm, never thought about this. What would be the use case for this?

joschi commented on graylog2-server/src/main/java/org/graylog2/alerts/types/FieldValueAlertCondition.java:
```diff
@@ -38,32 +43,85 @@
+                    "threshold_type",
+                    "Threshold Type",
+                    ThresholdType.HIGHER.toString(),
+                    Arrays.stream(ThresholdType.values()).collect(Collectors.toMap(Enum::toString, thresholdType -> thresholdType.toString().toLowerCase(Locale.ENGLISH))),
```
Since this pattern is used quite often (also in other entities), we should extract it into a common method.

Suggestion:
```java
public static Map<String, String> buildEnumMap(Class<? extends Enum> enumClass, Function<Enum, String> valueMapping) {
    return Arrays.stream(enumClass.getEnumConstants()).collect(Collectors.toMap(Enum::toString, valueMapping));
}

enum TestEnum {
    ONE, TWO
}

@Test
public void testBuildEnumMap() throws Exception {
    final Map<String, String> enumMap = buildEnumMap(TestEnum.class, (t) -> t.name().toLowerCase());
    assertThat(enumMap)
            .containsEntry("ONE", "one")
            .containsEntry("TWO", "two");
}
```

joschi commented on graylog2-server/src/test/java/org/graylog2/alerts/AlertConditionTest.java:
```diff
@@ -47,6 +47,18 @@
+    private abstract class TestAlertConditionFactory implements AlertCondition.Factory {
+        @Override
+        public AlertCondition.Config config() {
+            throw new RuntimeException("This method is not implements.");
```
Maybe [`UnsupportedOperationException`](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html) would be more fitting?

Also typo: "implements" &rarr; "implemented"

dennisoelkers replied:
✅

joschi commented on graylog2-server/src/test/java/org/graylog2/alerts/AlertConditionTest.java:
```diff
@@ -47,6 +47,18 @@
+
+        @Override
+        public AlertCondition.Descriptor descriptor() {
+            throw new RuntimeException("This method is not implements.");
```
See comment for `config()`,

dennisoelkers replied:
Looks suspiciously like there was some Copy & Paste coder at work.

dennisoelkers replied:
✅

joschi commented:
Creating a new alert condition in the web interface currently fails for me because `title` hasn't been set.

Example request sent to `StreamAlertConditionResource`:
```text
POST /api/streams/57f628ce4a13304252755c61/alerts/conditions HTTP/1.1
Host: 127.0.0.1:9000
Connection: keep-alive
Content-Length: 114
Pragma: no-cache
Cache-Control: no-cache
Authorization: Basic N2VlZmIyMzUtNTEwMi00MzM3LWJjMGItM2NjMTBmYmRiZjZlOnNlc3Npb24=
Origin: http://localhost:9000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.116 Safari/537.36
Content-Type: application/json
Accept: application/json
X-Requested-With: XMLHttpRequest
DNT: 1
Referer: http://localhost:9000/streams/57f628ce4a13304252755c61/alerts
Accept-Encoding: gzip, deflate
Accept-Language: de-DE,de;q=0.8,en-US;q=0.6,en;q=0.4

{"parameters":{"field":"test","time":1,"threshold":0,"threshold_type":"LOWER","type":"MEAN"},"type":"field_value"}
```

Exception:
```text
2016-10-06 12:43:25,901 ERROR: org.graylog2.shared.rest.exceptionmappers.AnyExceptionClassMapper - Unhandled exception in REST resource
java.lang.NullPointerException: null value in entry: title=null
	at com.google.common.collect.CollectPreconditions.checkEntryNotNull(CollectPreconditions.java:33) ~[guava-19.0.jar:?]
	at com.google.common.collect.ImmutableMapEntry.<init>(ImmutableMapEntry.java:50) ~[guava-19.0.jar:?]
	at com.google.common.collect.ImmutableMap.entryOf(ImmutableMap.java:122) ~[guava-19.0.jar:?]
	at com.google.common.collect.ImmutableMap$Builder.put(ImmutableMap.java:197) ~[guava-19.0.jar:?]
	at org.graylog2.alerts.AbstractAlertCondition.getPersistedFields(AbstractAlertCondition.java:135) ~[classes/:?]
	at org.graylog2.database.PersistedServiceImpl.embed(PersistedServiceImpl.java:255) ~[classes/:?]
	at org.graylog2.streams.StreamServiceImpl.addAlertCondition(StreamServiceImpl.java:267) ~[classes/:?]
	at org.graylog2.rest.resources.streams.alerts.StreamAlertConditionResource.create(StreamAlertConditionResource.java:98) ~[classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_102]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_102]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_102]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_102]
```

I have executed `mvn clean install` before, but I wouldn't rule out a local problem.

This being said, `org.graylog2.alerts.AbstractAlertCondition.getPersistedFields()` will definitely fail, if `title` is `null`.

dennisoelkers commented:
I fixed the issue with the missing title and added a mandatory title field to the alert condition form components. It was already included when there was no custom form component supplied and therefore the form was generated from the type definition.
